dvc/scm/git/backend/pygit2.py

```diff
@@ -223,7 +280,20 @@ def diff(self, rev_a: str, rev_b: str, binary=False) -> str:
         raise NotImplementedError
 
     def reset(self, hard: bool = False, paths: Iterable[str] = None):
-        raise NotImplementedError
+        from pygit2 import GIT_RESET_HARD, GIT_RESET_MIXED, IndexEntry
+
+        self.repo.index.read()
```

Comment:
`libgit2/pygit2` does not automatically write Git index file to disk in most cases, it's kept in-memory only until the repo object is closed. Since we are regularly swapping backends between git operations, we have to be careful to always explicitly read/write the `pygit2` to and from disk when doing any Git index related operations.